crates/kitsune2_showcase/src/app.rs

```diff
@@ -0,0 +1,132 @@
+use super::*;
+use bytes::Bytes;
+use kitsune2_api::*;
+use std::sync::Arc;
+
+// hard-coded random space
+const SPACE: SpaceId = SpaceId(Id(Bytes::from_static(&[
+    215, 33, 182, 196, 173, 34, 116, 214, 251, 139, 163, 71, 112, 51, 234, 167,
+    61, 62, 237, 27, 79, 162, 114, 232, 16, 184, 183, 235, 147, 138, 247, 202,
+])));
+
+pub struct App {
+    _k: DynKitsune,
+    s: DynSpace,
+    a: Arc<kitsune2_core::Ed25519LocalAgent>,
+    p: readline::Print,
+}
+
+impl App {
+    pub async fn new(print: readline::Print, args: Args) -> K2Result<Self> {
+        #[derive(Debug)]
+        struct S(readline::Print);
+
+        impl SpaceHandler for S {
+            fn recv_notify(
+                &self,
+                _to_agent: AgentId,
+                _from_agent: AgentId,
+                _space: SpaceId,
+                data: Bytes,
+            ) -> K2Result<()> {
+                let print = self.0.clone();
+                tokio::task::spawn(async move {
+                    print.print_line(String::from_utf8_lossy(&data).into());
+                });
+                Ok(())
+            }
+        }
+
+        #[derive(Debug)]
+        struct K(readline::Print);
+
+        impl KitsuneHandler for K {
+            fn create_space(
+                &self,
+                _space: SpaceId,
+            ) -> BoxFut<'_, K2Result<DynSpaceHandler>> {
+                Box::pin(async move {
+                    let s: DynSpaceHandler = Arc::new(S(self.0.clone()));
+                    Ok(s)
+                })
+            }
+
+            fn new_listening_address(
+                &self,
+                this_url: Url,
+            ) -> BoxFut<'static, ()> {
+                let print = self.0.clone();
+                Box::pin(async move {
+                    print.print_line(format!("Online at: {this_url}"));
```

Comment:
I see this printed twice on startup, with the same value. Is that expected?

Reply:
Yes... sortof. There is a race condition where the address could be bound before we hook up our notifier, so we also publish the new address manually, hence often getting two of these events. It's better to often get two than to sometimes get none. We can address that at a later date if we come up with a solution to it.